server/routes/nl_interface_next.py

```diff
@@ -592,50 +310,67 @@ def data():
     flask.abort(404)
   original_query = request.args.get('q')
   context_history = request.get_json().get('contextHistory', [])
-  has_context = False
-  if context_history:
-    has_context = True
   logging.info(context_history)
+
   query = str(escape(nl_utils.remove_punctuations(original_query)))
   embeddings_build = str(escape(request.args.get('build', "combined_all")))
-  default_place = "United States"
-  res = {'place_type': '', 'place_name': '', 'place_dcid': '', 'config': {}}
+  res = {
+      'place': {
+          'dcid': '',
+          'name': '',
+          'place_type': '',
+      },
+      'config': {},
+      'context': context_history
+  }
   if not query:
     logging.info("Query was empty")
     return _result_with_debug_info(res, "Aborted: Query was Empty.",
                                    embeddings_build,
-                                   _detection("", "", embeddings_build))
+                                   _detection("", "", embeddings_build),
+                                   context_history)
 
   # Query detection routine:
   # Returns detection for Place, SVs and Query Classifications.
-  recent_context = None
-  if context_history:
-    recent_context = context_history[-1]
   query_detection = _detection(str(escape(original_query)), query,
-                               embeddings_build, recent_context)
-
-  # Get Data Spec
-  data_spec = nl_data_spec.compute(query_detection)
-  page_config_pb = nl_page_config.build_page_config(query_detection, data_spec,
-                                                    context_history)
-  page_config = json.loads(MessageToJson(page_config_pb))
-
-  d = {
-      'place_type': query_detection.places_detected.main_place.place_type,
-      'place_name': query_detection.places_detected.main_place.name,
-      'place_dcid': query_detection.places_detected.main_place.dcid,
+                               embeddings_build)
+
+  # Generate new utterance.
+  prev_utterance = nl_utterance.load_utterance(context_history)
+  logging.info(prev_utterance)
+  utterance = nl_data_spec.compute(query_detection, prev_utterance)
+  logging.info(
+      nl_utterance.load_utterance(nl_utterance.save_utterance(utterance)))
+
+  if utterance.rankedCharts:
+    page_config_pb = nl_page_config.build_page_config(utterance)
+    page_config = json.loads(MessageToJson(page_config_pb))
+    # Use the first chart's place as main place.
+    main_place = utterance.rankedCharts[0].places[0]
+  else:
+    page_config = {}
+    main_place = Place(dcid='', name='', place_type='')
```

Comment:
should we also log here when there isn't a response?

Reply:
Done.